---
output: 
  github_document:
    toc: true
    toc_depth: 2
    fig_width: 6
    fig_height: 4
    dev: png
always_allow_html: true
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, 
                      collapse = TRUE,
                      comment = "#>",
                      fig.path = "man/figures/README-",
                      out.width = "100%", 
                      fig.retina = TRUE,
                      fig.width = 6,
                      fig.height = 4)
```

# `bayesVG`

<!-- badges: start -->
[![Language](https://img.shields.io/badge/-R?logo=R&logoColor=white)](https://github.com/topics/r)
![release](https://img.shields.io/github/v/release/jr-leary7/bayesVG?color=purple)
[![R-CMD-check](https://github.com/jr-leary7/bayesVG/actions/workflows/R-CMD-CHECK.yaml/badge.svg)](https://github.com/jr-leary7/bayesVG/actions/workflows/R-CMD-CHECK.yaml)
![last commit](https://img.shields.io/github/last-commit/jr-leary7/bayesVG/main?color=darkgreen)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Coverage](https://codecov.io/gh/jr-leary7/bayesVG/graph/badge.svg)](https://app.codecov.io/gh/jr-leary7/bayesVG)
[![CodeFactor](https://www.codefactor.io/repository/github/jr-leary7/bayesvg/badge/main)](https://www.codefactor.io/repository/github/jr-leary7/bayesvg/overview/main)
<!-- badges: end -->

# Installation

You can install the most recent version of `bayesVG` using:

```{r, eval=FALSE}
remotes::install_github("jr-leary7/bayesVG")
```

# Usage 

## Libraries 

```{r, eval=FALSE}
library(Seurat)
library(bayesVG)
```

## HVG detection 

### Data 

First, we load the 10X Genomics pbmc3k dataset, which is composed of 2,700 peripheral blood mononuclear cells from a single healthy donor. 

```{r, eval=FALSE}
data("seu_pbmc")
```

### Modeling 

Now we're able to model gene expression using variational inference, summarize the approximate posterior distribution of variance for each gene, and classify the top 3,000 most-variable genes as HVGs. The `findVariableFeaturesBayes()` function can take as input either a `Seurat` or a `SingleCellExperiment` object. 

```{r, eval=FALSE}
seu_pbmc <- findVariableFeaturesBayes(seu_pbmc, 
                                      n.cells.subsample = 500L, 
                                      algorithm = "meanfield",
                                      save.model = TRUE) %>% 
            classifyHVGs(n.HVG = 3000L)
```

We can generate the summary table (which is sorted by default) and extract the HVGs like so. These genes can then be used as the basis for downstream analyses such as PCA, clustering, UMAP visualization, etc. 

```{r, eval=FALSE}
summary_hvg <- getBayesianGeneStats(seu_pbmc)
top3k_hvgs <- summary_hvg$gene[1:3000]
```

The `extractModel()` function pulls out the fitted `brms` model from the object's unstructured metadata, allowing the user to perform posterior predictive checks (PPCs) and other diagnostics. Note: this requires that the HVG identification function be run with `save.model = TRUE`. 

```{r, eval=FALSE}
hvg_model <- extractModel(seu_pbmc)
```

## SVG detection 

### Data 

First, we load the 10X Genomics anterior mouse brain dataset that we've included in the package.

```{r, eval=FALSE}
data("seu_brain")
```

Before running `bayesVG` for SVG detection it's necessary to normalize the expression data and identify a set of 3,000 naive HVGs. 

```{r, eval=FALSE}
seu_brain <- NormalizeData(seu_brain, verbose = FALSE)
naive_hvgs <- getNaiveHVGs(seu_brain, n.hvg = 3000L)
```

### Modeling 

Now we can model gene expression with an approximate multivariate hierarchical Gaussian process (GP) via variational inference, summarize the spatial component of variance for each gene, and classify the top 1,000 most spatially variable genes as SVGs. The `findSpatiallyVariableFeaturesBayes()` function can take as input either a `Seurat` or a `SpatialExperiment` object. 

```{r, eval=FALSE}
seu_brain <- findSpatiallyVariableFeaturesBayes(seu_brain, 
                                                naive.hvgs = naive_hvgs, 
                                                likelihood = "gaussian", 
                                                kernel = "matern", 
                                                kernel.smoothness = 2.5, 
                                                algorithm = "meanfield", 
                                                n.cores = 4L, 
                                                save.model = TRUE) %>% 
             classifySVGs(n.SVG = 1000L)
```

We can extract the summary table (which like the HVG summary table is sorted by default)  and extract the top 1,000 SVGs like so. These genes can then be used as the basis for downstream analyses such as PCA, spatial clustering, UMAP visualization, etc. 

```{r, eval=FALSE}
summary_svg <- getBayesianGeneStats(seu_brain)
top1k_svgs <- summary_svg$gene[1:1000]
```

Since we ran the SVG identification model function with `save.model = TRUE`, we can extract the `cmdstanr` fit like so:

```{r, eval=FALSE}
svg_model <- extractModel(seu_brain)
```

We can cluster the SVG set (using a very fast modified Gaussian mixture model) into spatial modules as shown below. The clustering function returns a PCA embedding of the SVGs, a table of the soft cluster assignment probabilities, and the log-likelihood and Bayesian information criterion (BIC) of the clustering. The clustering code uses the fullrank VI algorithm by default due to the multi-modality and high correlation of the posterior distribution. 

```{r, eval=FALSE}
svg_clusters <- clusterSVGsBayes(seu_brain, 
                                 svgs = top1k_svgs, 
                                 n.clust = 5L, 
                                 n.cores = 4L)
```

Next, we can score the SVG clusters using [`UCell`](https://github.com/carmonalab/UCell) under the hood. These scores can then be visualized using e.g., violin plots or UMAPs. 

```{r, eval=FALSE}
seu_brain <- scoreSpatialModules(seu_brain, svg.clusters = svg_clusters)
```

Lastly, with our spatial modules identified and scores we can perform gene set enrichment analysis (GSEA) on each module; this is done internally using [`gProfiler2`](https://biit.cs.ut.ee/gprofiler/page/r). All you need to do is provide the results from `clusterSVGsBayes()` and specify the correct genus & species e.g., "hsapiens" for human or "mmusculus" for mouse. 

```{r, eval=FALSE}
enrich_res <- enrichSpatialModules(svg_clusters, species = "mmusculus")
```


# Contact information 

This package is developed & maintained by Jack R. Leary. Feel free to reach out by [opening an issue](https://github.com/jr-leary7/bayesVG/issues) or by email (j.leary@ufl.edu) if more detailed assistance is needed.

# References 

1. Jordan, M. *et al*. [An Introduction to Variational Methods for Graphical Models](https://doi.org/10.1023/A:1007665907178). *Machine Learning* (1999).

2. Burkner, P. [Bayesian Item Response Modeling in R with brms and Stan](https://doi.org/10.18637/jss.v100.i05). *Journal of Statistical Software* (2021).

3. Gabry, J. *et al*. [cmdstanr: R Interface to CmdStan](https://mc-stan.org/cmdstanr/). (2024).

4. Rasmussen, C. and Williams, C. [Gaussian Processes for Machine Learning](https://doi.org/10.7551/mitpress/3206.003.0019). *The MIT Press* (2005).

5. Riutort-Mayol, G. *et al*. [Practical Hilbert Space Approximate Bayesian Gaussian Processes for Probabilistic Programming](https://arxiv.org/abs/2004.11408). *arXiv* (2020).

6. Carpenter, B. *et al*. [Stan: A Probabilistic Programming Language](https://doi.org/10.18637/jss.v076.i01). *Journal of Statistical Software* (2017).
